xdem/volume.py

```diff
@@ -283,3 +284,135 @@ def hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_array]
     assert output is not None
 
     return output
+
+
+def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_array],
+                                    ref_dem: Union[np.ndarray, np.ma.masked_array],
+                                    mask: np.ndarray, min_coverage: float = 0.2,
+                                    plot: bool = False) -> np.ma.masked_array:
+    """
+    Interpolate a dDEM using local hypsometric interpolation.
+    The algorithm loops through each features in the vector file.
+
+    The dDEM is assumed to have been created as "voided_ddem = reference_dem - other_dem".
+
+    :param voided_ddem: A dDEM with voids (either an array with nans or a masked array).
+    :param ref_dem: The reference DEM in the dDEM comparison.
+    :param mask: A raster of same shape as voided_ddem and ref_dem, containing a diferent non-0 pixel value for \
+each geometry on which to loop.
+    :param min_coverage: The minimum coverage fraction to be considered for interpolation. Default is 0.2.
+    :param plot: Set to True to display intermediate plots.
+    """
+    # Remove any unnecessary dimension
+    orig_shape = voided_ddem.shape
+    voided_ddem = voided_ddem.squeeze()
+    ref_dem = ref_dem.squeeze()
+    mask = mask.squeeze()
+
+    # Check that all arrays have same dimensions
+    assert voided_ddem.shape == ref_dem.shape == mask.shape
+
+    # Get ddem array with invalid pixels converted to NaN and mask of invalid pixels
+    ddem, ddem_mask = xdem.spatial_tools.get_array_and_mask(voided_ddem)
+
+    # Get ref_dem array with invalid pixels converted to NaN and mask of invalid pixels
+    dem, dem_mask = xdem.spatial_tools.get_array_and_mask(ref_dem)
+
+    # A mask of inlier values: The union of the mask and the inverted exclusion masks of both rasters.
+    inlier_mask = (mask !=0) & (~ddem_mask & ~dem_mask)
+    if np.count_nonzero(inlier_mask) == 0:
+        warnings.warn("No valid data found within mask, returning copy", UserWarning)
+        return np.copy(ddem)
+
+    if plot:
+        plt.matshow(inlier_mask)
+        plt.title("inlier mask")
+        plt.show()
+
+    # List of indexes to loop on
+    geometry_index = np.unique(mask[mask!=0])
+    print("Found {:d} geometries".format(len(geometry_index)))
+
+    # Get fraction of valid pixels for each geometry
+    coverage = np.zeros(len(geometry_index))
+    for k, index in enumerate(geometry_index):
+        local_inlier_mask = inlier_mask & (mask == index)
+        total_pixels = np.count_nonzero((mask == index))
+        valid_pixels = np.count_nonzero(local_inlier_mask)
+        coverage[k] = valid_pixels/float(total_pixels)
+
+    # Filter geometries with too little coverage
+    valid_geometry_index = geometry_index[coverage >= min_coverage]
+    print("Found {:d} geometries with sufficient coverage".format(len(valid_geometry_index)))
+
+    idealized_ddem = -9999 * np.ones_like(dem)
+
+    for k, index in enumerate(valid_geometry_index):
```

Comment:
Could be nice in the future to experiment with multiprocessing/multithreading here.

Reply:
Added to issue #96